commit 83775ebb0bbf363d59fef1a7f9802d3981db6558
Author: Moritz Angermann <moritz.angermann@gmail.com>
Date:   Mon Oct 9 17:28:47 2023 +0800

    CgUtils.fixStgRegStmt respect register width
    
    This change ensure that the reg + offset computation is always of the
    same size.  Before this we could end up with a 64bit register, and then
    add a 32bit offset (on 32bit platforms).  This not only would fail type
    sanity checking, but also incorrectly truncate 64bit values into 32bit
    values silently on 32bit architectures.
    
    (cherry picked from commit dafc47091c9107dcf81e1e80a105f59211927c89)
    (cherry picked from commit 8e7a2065e433aa56552b335d420b5bf925082078)

Index: b/compiler/GHC/StgToCmm/CgUtils.hs
===================================================================
--- a/compiler/GHC/StgToCmm/CgUtils.hs
+++ b/compiler/GHC/StgToCmm/CgUtils.hs
@@ -173,15 +173,18 @@ fixStgRegStmt platform stmt = fixAssign
                         BaseReg -> baseAddr
                         _other  -> CmmLoad baseAddr (globalRegType platform reg) NaturallyAligned
 
-        CmmRegOff (CmmGlobal reg) offset ->
+        CmmRegOff greg@(CmmGlobal reg) offset ->
             -- RegOf leaves are just a shorthand form. If the reg maps
             -- to a real reg, we keep the shorthand, otherwise, we just
             -- expand it and defer to the above code.
+            -- NB: to ensure type correctness we need to ensure the Add
+            --     as well as the Int need to be of the same size as the
+            --     register.
             case reg `elem` activeStgRegs platform of
                 True  -> expr
-                False -> CmmMachOp (MO_Add (wordWidth platform)) [
-                                    fixExpr (CmmReg (CmmGlobal reg)),
+                False -> CmmMachOp (MO_Add (cmmRegWidth platform greg)) [
+                                    fixExpr (CmmReg greg),
                                     CmmLit (CmmInt (fromIntegral offset)
-                                                   (wordWidth platform))]
+                                                   (cmmRegWidth platform greg))]
 
         other_expr -> other_expr
